RabbitMQ.Stream.Client/PublishFilter.cs

```diff
@@ -0,0 +1,132 @@
+        private readonly List<(ulong, Message)> messages;
+        private readonly ILogger _logger;
+
+        private bool IsFilterSet()
+        {
+            return _filterValueExtractor != null;
+        }
+
+        private int MessageCount { get; }
+
+        public PublishFilter(byte publisherId, List<(ulong, Message)> messages,
+            Func<Message, string> filterValueExtractor, ILogger logger)
+        {
+            this.publisherId = publisherId;
+            this.messages = messages;
+            _filterValueExtractor = filterValueExtractor;
+            _logger = logger;
+            MessageCount = messages.Count;
+        }
+
+        public int Write(Span<byte> span)
+        {
+            var offset = WireFormatting.WriteUInt16(span, Key);
+            offset += WireFormatting.WriteUInt16(span[offset..], Version);
+            offset += WireFormatting.WriteByte(span[offset..], publisherId);
+            // this assumes we never write an empty publish frame
+            offset += WireFormatting.WriteInt32(span[offset..], MessageCount);
+            foreach (var (publishingId, msg) in messages)
+            {
+                try
+                {
+                    var filterValue = "";
+                    if (IsFilterSet())
+                    {
+                        // The try catch is mostly for the case where the filterValueExtractor
+                        // throws an exception. 
+                        // The user should be aware of this and make sure the function is safe
+                        // but in case of fail we have to skip the message
+                        filterValue = _filterValueExtractor(msg);
+                    }
+
+                    offset += WireFormatting.WriteUInt64(span[offset..], publishingId);
+                    if (IsFilterSet())
+                    {
+                        offset += WireFormatting.WriteString(span[offset..], filterValue);
+                    }
+                    else
+                    {
+                        offset += WireFormatting.WriteInt16(span[offset..], -1);
+                    }
+
+                    // this only write "simple" messages, we assume msg is just the binary body
+                    // not stream encoded data
+                    offset += WireFormatting.WriteUInt32(span[offset..], (uint)msg.Size);
+                    offset += msg.Write(span[offset..]);
+                }
+                catch (Exception e)
+                {
+                    // If there is an error on _filterValueExtractor we skip the message.
+                    // If there is an error on _filterValueExtractor the buffer here is still consistent.
+                    // so we can skip and continue
+                    _logger.LogError(e, "Error writing the filter message. Message won't be sent"
+                                        + "Check the filter value function");
+                }
+            }
```

Comment:
I _think_ the write function may not behave correctly in certain error scenarios. For example, it always writes the number of messages (let's say N number of messages). However, if there's an exception in `_filterValueExtractor()`, the message will be skipped. If we skip M number of messages, we will be sending `N-M` messages, however, we wrote that this frame has `N` messages. I'm not sure what's going to happen in the server with this frame 🤔 

I haven't checked what the Java implementation does, but I would consider re-throwing and logging an error somewhere, probably in the caller of `write`

Reply:
[Here](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/blob/stream-filtering/RabbitMQ.Stream.Client/PublishFilter.cs#L36-L44) there is the fix for the message count. 

The size calculation happens before sending the messages so we can decrease the message count to send. 

Read the internal conversation about the [error handling](https://vmware.slack.com/archives/CSC92F50B/p1687180929189639)  in the case of Filter. The same problem happens when we consume the message for the post filter.  See [here](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/blob/stream-filtering/RabbitMQ.Stream.Client/RawConsumer.cs#L251-L255)